bin/test-workflow/8_app_verify_authentication.sh

```diff
@@ -0,0 +1,183 @@
+function finish {
+  readonly PIDS=(
+    "SIDECAR_PORT_FORWARD_PID"
+    "INIT_PORT_FORWARD_PID"
+    "INIT_WITH_HOST_OUTSIDE_APPS_PORT_FORWARD_PID"
+    "SECRETLESS_PORT_FORWARD_PID"
+  )
+
+  if [[ "$DETAILED_DUMP_ON_EXIT" == "true" ]]; then
+    dump_kubernetes_resources
+    dump_authentication_policy
+  fi
+
+  set +u
+
+  echo -e "\n\nStopping all port-forwarding"
+  for pid in "${PIDS[@]}"; do
+    if [ -n "${!pid}" ]; then
+      # Kill process, and swallow any errors
+      kill "${!pid}" > /dev/null 2>&1
+    fi
+  done
+}
+trap finish EXIT
+
+announce "Validating that the deployments are functioning as expected."
+
+set_namespace "$TEST_APP_NAMESPACE_NAME"
+
+deploy_test_curl() {
+  $cli delete --ignore-not-found pod/test-curl
+  $cli create -f ./$PLATFORM/test-curl.yml
+}
+
+check_test_curl() {
+  pods_ready "test-curl"
+}
+
+pod_curl() {
+  kubectl exec test-curl -- curl "$@"
+}
+
+if [[ "$TEST_APP_LOADBALANCER_SVCS" == "false" ]]; then
+  echo "Deploying a test curl pod"
+  deploy_test_curl
+  echo "Waiting for test curl pod to become available"
+  bl_retry_constant "${RETRIES}" "${RETRY_WAIT}"  check_test_curl
+fi
+  
+echo "Waiting for pods to become available"
+
+check_pods(){
+  # pods_ready "test-app-summon-init" &&
+  # pods_ready "test-app-with-host-outside-apps-branch-summon-init" &&
+  pods_ready "test-app-summon-sidecar" # &&
+  # pods_ready "test-app-secretless"
+}
+bl_retry_constant "${RETRIES}" "${RETRY_WAIT}"  check_pods
+
+if [[ "$PLATFORM" == "openshift" ]]; then
+  echo "Waiting for deployments to become available"
+
+  check_deployment_status(){
+    # [[ "$(deployment_status "test-app-summon-init")" == "Complete" ]] &&
+    # [[ "$(deployment_status "test-app-with-host-outside-apps-branch-summon-init")" == "Complete" ]] &&
+    [[ "$(deployment_status "test-app-summon-sidecar")" == "Complete" ]] # &&
+    # [[ "$(deployment_status "test-app-secretless")" == "Complete" ]]
+  }
+  bl_retry_constant "${RETRIES}" "${RETRY_WAIT}"  check_deployment_status
+
+  sidecar_pod=$(get_pod_name test-app-summon-sidecar)
+  init_pod=$(get_pod_name test-app-summon-init)
+  init_pod_with_host_outside_apps=$(get_pod_name test-app-with-host-outside-apps-branch-summon-init)
+  secretless_pod=$(get_pod_name test-app-secretless)
+
+  # Routes are defined, but we need to do port-mapping to access them
+  oc port-forward "$sidecar_pod" 8081:8080 > /dev/null 2>&1 &
+  SIDECAR_PORT_FORWARD_PID=$!
+  oc port-forward "$init_pod" 8082:8080 > /dev/null 2>&1 &
+  INIT_PORT_FORWARD_PID=$!
+  oc port-forward "$secretless_pod" 8083:8080 > /dev/null 2>&1 &
+  SECRETLESS_PORT_FORWARD_PID=$!
+  oc port-forward "$init_pod_with_host_outside_apps" 8084:8080 > /dev/null 2>&1 &
+  INIT_WITH_HOST_OUTSIDE_APPS_PORT_FORWARD_PID=$!
+
+  curl_cmd=curl
+  sidecar_url="localhost:8081"
+  init_url="localhost:8082"
+  secretless_url="localhost:8083"
+  init_url_with_host_outside_apps="localhost:8084"
+else
+  if [[ "$TEST_APP_LOADBALANCER_SVCS" == "true" ]]; then
```

Comment:
Maybe something to discuss this in a group meeting.... We should consider whether testing our applications using loadbalancers is giving us any real value. It seems to me that using the loadbalancer external IPs to test our applications' services versus just using a Kubernetes internal service name is testing the loadbalancer functionality of whatever Kubernetes platform we're using, not Conjur functionality.

Dropping the requirement that we need to test using external IPs would greatly simplify the testing here. For example, if we ONLY test using internal service names, then everything can be tested from inside the Kubernetes cluster (using a `curl Pod`).

This would mean that we might be able to implement the verification tests for each application integration separately, in their own Helm test. AND, it may be that running `helm test ....` on a parent chart does a Helm test on all of its subcharts????? I haven't found documentation on this.... but even if this isn't the case, we should be able to run `helm test ...` on each subchart individually.

The bottom line is that each subchart would be able to provide its own test mechanism.

Reply:
Removed the loadbalancer testing option based on discussion